config/v3_0/types/config.go

```diff
@@ -24,3 +30,49 @@ var (
 		Minor: 0,
 	}
 )
+
+func (cfg Config) Validate(c path.ContextPath) (r report.Report) {
+	systemdPath := "/etc/systemd/system/"
+	unit_paths := make([]string, 0)
+	for _, unit := range cfg.Systemd.Units {
+		if unit.Contents != nil {
+			path_string := systemdPath + unit.Name
+			unit_paths = append(unit_paths, path_string)
+		} else {
+			for _, dropin := range unit.Dropins {
+				if dropin.Contents != nil {
+					path_string := systemdPath + unit.Name + ".d/" + dropin.Name
+					unit_paths = append(unit_paths, path_string)
+				}
+			}
+		}
+	}
+	for i, f := range cfg.Storage.Files {
+		if validateSystemdConflict(f.Path, unit_paths) {
+			r.AddOnError(c.Append("storage", "files", i, "path"), errors.ErrPathConflictsSystemd)
+		}
+	}
+	for i, d := range cfg.Storage.Directories {
+		if validateSystemdConflict(d.Path, unit_paths) {
+			r.AddOnError(c.Append("storage", "directories", i, "path"), errors.ErrPathConflictsSystemd)
+		}
+	}
+	for i, l := range cfg.Storage.Links {
+		if validateSystemdConflict(l.Path, unit_paths) {
+			r.AddOnError(c.Append("storage", "links", i, "path"), errors.ErrPathConflictsSystemd)
+		}
+	}
+	return
+}
+
+func validateSystemdConflict(p string, unit_paths []string) bool {
+	systemdPath := "/etc/systemd/system/"
+	if strings.HasPrefix(p, systemdPath) {
```

Comment:
Once `unitPaths` is a set, lookups will be constant time, and we can drop this check.  At that point this function will reduce to one line and should just be merged into the caller.